File: Hashing/HashTableOperations.java
// File Name: HashTableOperations.java

import java.util.Scanner;

class HashTableOperations {

    private Integer[] table;
    private int size;
    private final Integer DELETED = Integer.MIN_VALUE;

    public HashTableOperations(int size) {
        this.size = size;
        table = new Integer[size];
    }

    private int hashFunction(int key) {
        return key % size;
    }

    public void insert(int key) {
        int index = hashFunction(key);
        int start = index;

        while (table[index] != null && table[index] != DELETED) {
            index = (index + 1) % size;
            if (index == start) {
                System.out.println("Hash Table is Full");
                return;
            }
        }

        table[index] = key;
        System.out.println("Key inserted at index " + index);
    }

    public void search(int key) {
        int index = hashFunction(key);
        int start = index;

        while (table[index] != null) {
            if (table[index] != DELETED && table[index] == key) {
                System.out.println("Key found at index " + index);
                return;
            }
            index = (index + 1) % size;
            if (index == start)
                break;
        }

        System.out.println("Key not found");
    }

    public void delete(int key) {
        int index = hashFunction(key);
        int start = index;

        while (table[index] != null) {
            if (table[index] != DELETED && table[index] == key) {
                table[index] = DELETED;
                System.out.println("Key deleted");
                return;
            }
            index = (index + 1) % size;
            if (index == start)
                break;
        }

        System.out.println("Key not found");
    }

    public void display() {
        System.out.println("\nHash Table:");
        for (int i = 0; i < size; i++) {
            if (table[i] == null)
                System.out.println("Index " + i + " -> EMPTY");
            else if (table[i] == DELETED)
                System.out.println("Index " + i + " -> DELETED");
            else
                System.out.println("Index " + i + " -> " + table[i]);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter size: ");
        int size = sc.nextInt();

        HashTableOperations ht = new HashTableOperations(size);

        while (true) {
            System.out.println("\n1.Insert\n2.Search\n3.Delete\n4.Display\n5.Exit");
            int choice = sc.nextInt();

            switch (choice) {
                case 1:
                    ht.insert(sc.nextInt());
                    break;
                case 2:
                    ht.search(sc.nextInt());
                    break;
                case 3:
                    ht.delete(sc.nextInt());
                    break;
                case 4:
                    ht.display();
                    break;
                case 5:
                    System.exit(0);
            }
        }
    }
}

File: Hashing/hash_table_operations.c
/* File Name: hash_table_operations.c */

#include <stdio.h>
#include <stdlib.h>

#define SIZE 10
#define EMPTY -1
#define DELETED -2

int hashTable[SIZE];

// Initialize table
void initialize() {
    for (int i = 0; i < SIZE; i++)
        hashTable[i] = EMPTY;
}

// Hash Function
int hashFunction(int key) {
    return key % SIZE;
}

// Insert Operation
void insert(int key) {
    int index = hashFunction(key);
    int start = index;

    while (hashTable[index] != EMPTY && hashTable[index] != DELETED) {
        index = (index + 1) % SIZE;
        if (index == start) {
            printf("Hash Table is Full\n");
            return;
        }
    }

    hashTable[index] = key;
    printf("Key %d inserted at index %d\n", key, index);
}

// Search Operation
void search(int key) {
    int index = hashFunction(key);
    int start = index;

    while (hashTable[index] != EMPTY) {
        if (hashTable[index] == key) {
            printf("Key %d found at index %d\n", key, index);
            return;
        }
        index = (index + 1) % SIZE;
        if (index == start)
            break;
    }

    printf("Key %d not found\n", key);
}

// Delete Operation
void deleteKey(int key) {
    int index = hashFunction(key);
    int start = index;

    while (hashTable[index] != EMPTY) {
        if (hashTable[index] == key) {
            hashTable[index] = DELETED;
            printf("Key %d deleted\n", key);
            return;
        }
        index = (index + 1) % SIZE;
        if (index == start)
            break;
    }

    printf("Key %d not found\n", key);
}

// Display
void display() {
    printf("\nHash Table:\n");
    for (int i = 0; i < SIZE; i++) {
        if (hashTable[i] == EMPTY)
            printf("Index %d -> EMPTY\n", i);
        else if (hashTable[i] == DELETED)
            printf("Index %d -> DELETED\n", i);
        else
            printf("Index %d -> %d\n", i, hashTable[i]);
    }
}

int main() {
    int choice, key;
    initialize();

    while (1) {
        printf("\n1.Insert\n2.Search\n3.Delete\n4.Display\n5.Exit\n");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter key: ");
                scanf("%d", &key);
                insert(key);
                break;
            case 2:
                printf("Enter key: ");
                scanf("%d", &key);
                search(key);
                break;
            case 3:
                printf("Enter key: ");
                scanf("%d", &key);
                deleteKey(key);
                break;
            case 4:
                display();
                break;
            case 5:
                exit(0);
        }
    }
}
